src/sage/rings/padics/padic_valuation.py

```diff
@@ -0,0 +1,1319 @@
+          result in the return value.
+
+        - ``assume_squarefree`` -- a boolean (default: ``False``); whether to
+          assume that ``G`` is square-free over the completion of the domain of
+          this valuation. Setting this to ``True`` can significantly improve
+          the performance.
+
+        EXAMPLES:
+
+        We consider an extension as unramified if its ramification index is 1.
+        Hence, a trivial extension is unramified::
+
+            sage: R.<x> = QQ[]
+            sage: v = QQ.valuation(2)
+            sage: v.is_unramified(x)
+            True
+
+        If ``G`` remains irreducible in reduction, then it defines an
+        unramified extension::
+
+            sage: v.is_unramified(x^2 + x + 1)
+            True
+
+        However, even if ``G`` factors, it might define an unramified
+        extension::
+
+            sage: v.is_unramified(x^2 + 2*x + 4)
+            True
+
+        """
+        R = G.parent()
+
+        from sage.rings.polynomial.polynomial_ring import is_PolynomialRing
+        if not is_PolynomialRing(R) or R.base_ring() is not self.domain() or not len(R.gens()) == 1 or not G.is_monic():
+            raise ValueError("G must be a monic univariate polynomial over the domain of this valuation")
+        if not assume_squarefree and not G.is_squarefree():
+            raise ValueError("G must be squarefree")
+
+        from sage.rings.valuation.gauss_valuation import GaussValuation
+
+        steps = [ GaussValuation(R, self) ]
+        while True:
+            v = steps[-1]
+            if v.E() > 1:
+                ret = False
+                break
+            if v.F() == G.degree():
+                ret = True
+                break
+
+            assert v(G) is not infinity
+            if v.is_key(G):
+                ret = True
+                break
+
+            next = v.mac_lane_step(G, assume_squarefree=True)
+            if len(next)>1:
+                ret = False
+                break
+            steps.append(next[0])
+
+        if include_steps:
+            return ret, steps
+        else:
+            return ret
+
+    def is_totally_ramified(self, G, include_steps=False, assume_squarefree=False):
+        """
+        Return whether ``G`` defines a single totally ramified extension of the
+        completion of the domain of this valuation.
+
+        INPUT:
+
+        - ``G`` -- a monic squarefree polynomial over the domain of this valuation
+
+        - ``include_steps`` -- a boolean (default: ``False``); where to include
+          the valuations produced during the process of checking whether ``G``
+          is totally ramified in the return value
+
+        - ``assume_squarefree`` -- a boolean (default: ``False``); whether to
+          assume that ``G`` is square-free over the completion of the domain of
+          this valuation. Setting this to ``True`` can significantly improve
+          the performance.
+
```

Comment:
It would be good to have a remark about whether trivial valuations are totally ramified.